File: .explorer-v1-backups/20260907-223640/src/competition_navigation/scripts/corridor_graph.py
#!/usr/bin/env python3
"""Small, deterministic graph model for floor exploration tasks.

The graph stores only nodes discovered by navigation and does not depend on
the building metadata.  Motion planning remains the responsibility of the
navigation node; this module only handles topology and task ordering.
"""

from __future__ import print_function

import math


class GraphNode(object):
    """A discovered navigation node."""

    def __init__(self, node_id, node_type, floor, x, y, yaw=0.0,
                 corridor_s=None, confidence=0.0, label=""):
        self.node_id = int(node_id)
        self.node_type = str(node_type)
        self.floor = int(floor)
        self.x = float(x)
        self.y = float(y)
        self.yaw = float(yaw)
        self.corridor_s = None if corridor_s is None else float(corridor_s)
        self.confidence = float(confidence)
        self.label = str(label)
        self.visited = False
        self.entered = False
        self.interior_observation = False
        self.coverage = 0.0
        self.completed = False
        self.blocked = False

    @property
    def position(self):
        return self.x, self.y

    def distance_to(self, x, y):
        return math.hypot(self.x - float(x), self.y - float(y))

    def as_dict(self):
        return {
            "id": self.node_id,
            "type": self.node_type,
            "floor": self.floor,
            "x": self.x,
            "y": self.y,
            "yaw": self.yaw,
            "corridor_s": self.corridor_s,
            "confidence": self.confidence,
            "label": self.label,
            "visited": self.visited,
            "entered": self.entered,
            "interior_observation": self.interior_observation,
            "coverage": self.coverage,
            "completed": self.completed,
            "blocked": self.blocked,
        }


class GraphEdge(object):
    """An undirected traversable connection between two nodes."""

    def __init__(self, edge_id, start, end, length=None):
        self.edge_id = int(edge_id)
        self.start = int(start)
        self.end = int(end)
        self.length = None if length is None else float(length)
        self.visited = False
        self.blocked = False

    def other(self, node_id):
        node_id = int(node_id)
        if node_id == self.start:
            return self.end
        if node_id == self.end:
            return self.start
        raise ValueError("node %d is not incident to edge %d" %
                         (node_id, self.edge_id))


class CorridorGraph(object):
    """Mutable floor graph with deterministic DFS task ordering."""

    def __init__(self, merge_distance=0.5):
        self.merge_distance = float(merge_distance)
        self.nodes = {}
        self.edges = {}
        self._next_node_id = 0
        self._next_edge_id = 0

    def add_node(self, node_type, floor, x, y, yaw=0.0, corridor_s=None,
                 confidence=0.0, label="", merge=True):
        """Add a node, merging nearby same-floor observations when requested."""
        if merge:
            existing = self._nearest_node(node_type, floor, x, y)
            if existing is not None:
                existing.confidence = max(existing.confidence, float(confidence))
                if existing.corridor_s is None and corridor_s is not None:
                    existing.corridor_s = float(corridor_s)
                if label and not existing.label:
                    existing.label = str(label)
                return existing.node_id, False

        node = GraphNode(
            self._next_node_id,
            node_type,
            floor,
            x,
            y,
            yaw,
            corridor_s,
            confidence,
            label,
        )
        self.nodes[node.node_id] = node
        self._next_node_id += 1
        return node.node_id, True

    def add_room_entry(self, floor, x, y, yaw=0.0, corridor_s=None,
                       confidence=0.0, label=""):
        return self.add_node(
            "room_entry", floor, x, y, yaw, corridor_s, confidence, label
        )

    def add_edge(self, start, end, length=None):
        """Add an undirected edge and return its stable id."""
        start, end = int(start), int(end)
        if start == end:
            raise ValueError("self-loop is not a corridor edge")
        if start not in self.nodes or end not in self.nodes:
            raise KeyError("edge endpoint is not in graph")
        for edge in self.edges.values():
            if {edge.start, edge.end} == {start, end}:
                return edge.edge_id, False
        if length is None:
            length = math.hypot(
                self.nodes[start].x - self.nodes[end].x,
                self.nodes[start].y - self.nodes[end].y,
            )
        edge = GraphEdge(self._next_edge_id, start, end, length)
        self.edges[edge.edge_id] = edge
        self._next_edge_id += 1
        return edge.edge_id, True

    def mark_node(self, node_id, visited=None, completed=None, blocked=None):
        node = self.nodes[int(node_id)]
        if visited is not None:
            node.visited = bool(visited)
        if completed is not None:
            node.completed = bool(completed)
        if blocked is not None:
            node.blocked = bool(blocked)

    def mark_edge(self, edge_id, visited=None, blocked=None):
        edge = self.edges[int(edge_id)]
        if visited is not None:
            edge.visited = bool(visited)
        if blocked is not None:
            edge.blocked = bool(blocked)

    def neighbors(self, node_id, include_blocked=False):
        node_id = int(node_id)
        result = []
        for edge in self.edges.values():
            if edge.blocked and not include_blocked:
                continue
            if edge.start == node_id or edge.end == node_id:
                result.append((edge, self.nodes[edge.other(node_id)]))
        return result

    def pending_rooms(self, floor, descending=True):
        rooms = [
            node for node in self.nodes.values()
            if (node.floor == int(floor) and node.node_type == "room_entry" and
                not node.completed and not node.blocked)
        ]
        return sorted(
            rooms,
            key=lambda node: (
                float("-inf") if node.corridor_s is None else node.corridor_s,
                node.node_id,
            ),
            reverse=bool(descending),
        )

    def dfs_order(self, start_node_id, floor=None):
        """Return a deterministic DFS node order without changing node state."""
        start_node_id = int(start_node_id)
        if start_node_id not in self.nodes:
            raise KeyError("DFS start node is not in graph")
        visited = set()
        order = []

        def sort_key(item):
            edge, node = item
            type_priority = {
                "room_entry": 0,
                "corridor_end": 1,
                "junction": 2,
                "stair": 3,
                "entrance": 4,
            }.get(node.node_type, 5)
            distance = (
                float("-inf") if node.corridor_s is None else node.corridor_s
            )
            return type_priority, -distance, edge.edge_id

        def visit(node_id):
            if node_id in visited:
                return
            visited.add(node_id)
            node = self.nodes[node_id]
            if floor is None or node.floor == int(floor):
                order.append(node_id)
            for edge, neighbor in sorted(self.neighbors(node_id), key=sort_key):
                if neighbor.node_id in visited:
                    continue
                if floor is not None and neighbor.floor != int(floor):
                    continue
                edge.visited = True
                visit(neighbor.node_id)

        visit(start_node_id)
        return order

    def _nearest_node(self, node_type, floor, x, y):
        candidates = [
            node for node in self.nodes.values()
            if node.node_type == str(node_type) and node.floor == int(floor)
        ]
        if not candidates:
            return None
        nearest = min(candidates, key=lambda node: node.distance_to(x, y))
        return nearest if nearest.distance_to(x, y) <= self.merge_distance else None

File: .explorer-v1-backups/20260907-232950/src/competition_navigation_fastlio2_cpp/src/corridor_graph.cpp
// C++ port of corridor_graph.py.  See corridor_graph.h for the contract.
#include "competition_navigation_fastlio2_cpp/corridor_graph.h"

#include <algorithm>
#include <cmath>
#include <functional>
#include <limits>

namespace competition_navigation_fastlio2_cpp {

CorridorGraph::CorridorGraph(double merge_distance)
    : merge_distance_(merge_distance) {}

std::pair<int, bool> CorridorGraph::add_node(
    const std::string& node_type, int floor, double x, double y, double yaw,
    std::optional<double> corridor_s, double confidence,
    const std::string& label, bool merge) {
  if (merge) {
    int existing = nearest_node(node_type, floor, x, y);
    if (existing >= 0) {
      GraphNode& node = nodes.at(existing);
      node.confidence = std::max(node.confidence, confidence);
      if (!node.corridor_s.has_value() && corridor_s.has_value()) {
        node.corridor_s = corridor_s;
      }
      if (!label.empty() && node.label.empty()) {
        node.label = label;
      }
      return {node.node_id, false};
    }
  }

  GraphNode node;
  node.node_id = next_node_id_++;
  node.node_type = node_type;
  node.floor = floor;
  node.x = x;
  node.y = y;
  node.yaw = yaw;
  node.corridor_s = corridor_s;
  node.confidence = confidence;
  node.label = label;
  nodes[node.node_id] = node;
  return {node.node_id, true};
}

std::pair<int, bool> CorridorGraph::add_room_entry(
    int floor, double x, double y, double yaw, std::optional<double> corridor_s,
    double confidence, const std::string& label) {
  return add_node("room_entry", floor, x, y, yaw, corridor_s, confidence, label);
}

std::pair<int, bool> CorridorGraph::add_edge(int start, int end,
                                             std::optional<double> length) {
  if (start == end) {
    // Python raised ValueError here; navigation never creates a self-loop.
    return {-1, false};
  }
  if (nodes.find(start) == nodes.end() || nodes.find(end) == nodes.end()) {
    return {-1, false};
  }
  for (const auto& kv : edges) {
    const GraphEdge& edge = kv.second;
    if ((edge.start == start && edge.end == end) ||
        (edge.start == end && edge.end == start)) {
      return {edge.edge_id, false};
    }
  }
  double resolved = length.has_value()
                        ? length.value()
                        : std::hypot(nodes.at(start).x - nodes.at(end).x,
                                     nodes.at(start).y - nodes.at(end).y);
  GraphEdge edge;
  edge.edge_id = next_edge_id_++;
  edge.start = start;
  edge.end = end;
  edge.length = resolved;
  edges[edge.edge_id] = edge;
  return {edge.edge_id, true};
}

void CorridorGraph::mark_node(int node_id, std::optional<bool> visited,
                              std::optional<bool> completed,
                              std::optional<bool> blocked) {
  GraphNode& node = nodes.at(node_id);
  if (visited.has_value()) node.visited = visited.value();
  if (completed.has_value()) node.completed = completed.value();
  if (blocked.has_value()) node.blocked = blocked.value();
}

void CorridorGraph::mark_edge(int edge_id, std::optional<bool> visited,
                              std::optional<bool> blocked) {
  GraphEdge& edge = edges.at(edge_id);
  if (visited.has_value()) edge.visited = visited.value();
  if (blocked.has_value()) edge.blocked = blocked.value();
}

std::vector<std::pair<GraphEdge, GraphNode>> CorridorGraph::neighbors(
    int node_id, bool include_blocked) const {
  std::vector<std::pair<GraphEdge, GraphNode>> result;
  for (const auto& kv : edges) {
    const GraphEdge& edge = kv.second;
    if (edge.blocked && !include_blocked) continue;
    if (edge.start == node_id || edge.end == node_id) {
      result.emplace_back(edge, nodes.at(edge.other(node_id)));
    }
  }
  return result;
}

std::vector<GraphNode> CorridorGraph::pending_rooms(int floor,
                                                   bool descending) const {
  std::vector<GraphNode> rooms;
  for (const auto& kv : nodes) {
    const GraphNode& node = kv.second;
    if (node.floor == floor && node.node_type == "room_entry" &&
        !node.completed && !node.blocked) {
      rooms.push_back(node);
    }
  }
  std::sort(rooms.begin(), rooms.end(),
            [](const GraphNode& a, const GraphNode& b) {
              double a_s = a.corridor_s.has_value()
                               ? a.corridor_s.value()
                               : -std::numeric_limits<double>::infinity();
              double b_s = b.corridor_s.has_value()
                               ? b.corridor_s.value()
                               : -std::numeric_limits<double>::infinity();
              if (a_s != b_s) return a_s < b_s;
              return a.node_id < b.node_id;
            });
  if (descending) std::reverse(rooms.begin(), rooms.end());
  return rooms;
}

std::vector<int> CorridorGraph::dfs_order(int start_node_id,
                                          std::optional<int> floor) {
  std::vector<int> order;
  std::map<int, bool> visited;

  // Matches the Python sort_key: (type_priority, -distance, edge_id).
  auto type_priority = [](const std::string& t) {
    if (t == "room_entry") return 0;
    if (t == "corridor_end") return 1;
    if (t == "junction") return 2;
    if (t == "stair") return 3;
    if (t == "entrance") return 4;
    return 5;
  };

  std::function<void(int)> visit = [&](int node_id) {
    if (visited[node_id]) return;
    visited[node_id] = true;
    const GraphNode& node = nodes.at(node_id);
    if (!floor.has_value() || node.floor == floor.value()) {
      order.push_back(node_id);
    }
    // Collect incident (edge_id, other_node) pairs, sorted by sort_key.
    std::vector<std::pair<int, int>> incident;  // (edge_id, other_node_id)
    for (const auto& kv : edges) {
      const GraphEdge& edge = kv.second;
      if (edge.blocked) continue;
      if (edge.start == node_id || edge.end == node_id) {
        incident.emplace_back(edge.edge_id, edge.other(node_id));
      }
    }
    std::sort(incident.begin(), incident.end(),
              [&](const std::pair<int, int>& a, const std::pair<int, int>& b) {
                const GraphNode& na = nodes.at(a.second);
                const GraphNode& nb = nodes.at(b.second);
                int pa = type_priority(na.node_type);
                int pb = type_priority(nb.node_type);
                if (pa != pb) return pa < pb;
                double da = na.corridor_s.has_value()
                                ? na.corridor_s.value()
                                : -std::numeric_limits<double>::infinity();
                double db = nb.corridor_s.has_value()
                                ? nb.corridor_s.value()
                                : -std::numeric_limits<double>::infinity();
                if (da != db) return -da < -db;  // -distance ascending
                return a.first < b.first;        // edge_id
              });
    for (const auto& inc : incident) {
      int edge_id = inc.first;
      int neighbor_id = inc.second;
      if (visited.find(neighbor_id) != visited.end()) continue;
      if (floor.has_value() && nodes.at(neighbor_id).floor != floor.value())
        continue;
      edges.at(edge_id).visited = true;
      visit(neighbor_id);
    }
  };

  if (nodes.find(start_node_id) == nodes.end()) {
    return order;  // Python raised KeyError; navigation passes a valid id.
  }
  visit(start_node_id);
  return order;
}

int CorridorGraph::nearest_node(const std::string& node_type, int floor,
                                double x, double y) const {
  int nearest = -1;
  double nearest_dist = 0.0;
  for (const auto& kv : nodes) {
    const GraphNode& node = kv.second;
    if (node.node_type != node_type || node.floor != floor) continue;
    double d = node.distance_to(x, y);
    if (nearest < 0 || d < nearest_dist) {
      nearest = node.node_id;
      nearest_dist = d;
    }
  }
  if (nearest < 0) return -1;
  return nearest_dist <= merge_distance_ ? nearest : -1;
}

}  // namespace competition_navigation_fastlio2_cpp
